### PR TITLE
Parts 2 a, b ,c by Artem

### PR DESCRIPTION
**Pull request by Atom0713**

### CODE REVIEW

File: Assigment_1/Assignment 1.ipynb [hunk 1]
```diff
@@ -2422,19 +2452,53 @@
...
+    "        else:\n",
+    "            total_auc += (cov_tp/tot_tp)*(fp_i/tot_fp) + (tp_i/tot_tp)*(fp_i/tot_fp)/2\n",
+    "            cov_tp += tp_i\n",
+    "   \n",
+    "    \n",
+    "    return total_auc\n",
+    "\n",
+    "\n",
     "\n"
    ]
```
Comment (Eugenius0) [inline Assigment_1/Assignment 1.ipynb:2490]: Here I have a slightly modified version: Updated the auc function to calculate AUC for each class individually and then combine them using class frequency weights for a correct overall AUC score. Let me know what you think :)

def auc(df, correctlabels):
    aucs = {}
    class_counts = pd.Series(correctlabels).value_counts(normalize=True)
    
    for column in df.columns:
        scores = {}
        
        for correctlabel, probability in zip(correctlabels, df[column]):
            is_true_positive = 1 if correctlabel == column else 0
            score = [is_true_positive, int(not is_true_positive)]
            if probability in scores:
                scores[probability] = [x + y for x, y in zip(scores[probability], score)]
            else:
                scores[probability] = score

        sorted_scores = dict(sorted(scores.items(), reverse=True))
        
        tot_tp = sum([tp[0] for tp in sorted_scores.values()])
        tot_fp = sum([tp[1] for tp in sorted_scores.values()])
        cov_tp = 0
        class_auc = 0
        for _, score in sorted_scores.items():
            tp_i, fp_i = score
            if fp_i == 0:
                cov_tp += tp_i
            elif tp_i == 0:
                class_auc += (cov_tp / tot_tp) * (fp_i / tot_fp)
            else:
                class_auc += (cov_tp / tot_tp) * (fp_i / tot_fp) + (tp_i / tot_tp) * (fp_i / tot_fp) / 2
                cov_tp += tp_i

        aucs[column] = class_auc
    
    weighted_auc = sum(aucs[class_label] * class_counts[class_label] for class_label in aucs)
    
    return weighted_auc

File: Assigment_1/Assignment 1.ipynb [hunk 2]
```diff
@@ -2284,12 +2284,32 @@
...
+    "\n",
+    "import math\n",
+    "\n",
+    "def folds(df, nofolds=10):\n",
+    "    new_df = df.copy()\n",
+    "    folds = []\n",
+    "\n",
+    "    # Shuffle the DataFrame indices\n",
+    "    shuffled_indices = np.random.permutation(new_df.index)\n",
+    "    fold_size = math.ceil(len(new_df) / nofolds)\n",
```
[the remark this answers is not in the document]
Comment (Atom0713) [reply Assigment_1/Assignment 1.ipynb:2296]: Floor division does not provide equal sized folds

image

File: Assigment_1/Assignment 1.ipynb [hunk 3]
```diff
@@ -2284,12 +2284,32 @@
...
+    "\n",
+    "    for i in range(nofolds):\n",
```
[the remark this answers is not in the document]
Comment (Atom0713) [reply Assigment_1/Assignment 1.ipynb:2298]: I can do that.

File: Assigment_1/Assignment 1.ipynb [hunk 4]
```diff
@@ -2284,12 +2284,32 @@
...
+    "        end_idx = (i + 1) * fold_size if i != nofolds - 1 else len(new_df)\n",
```
[the remark this answers is not in the document]
Comment (Atom0713) [reply Assigment_1/Assignment 1.ipynb:2299]: Not sure we can. Does `(i + 1) == start_index == i * fold_size` ?

File: Assigment_1/Assignment 1.ipynb [hunk 5]
```diff
@@ -2347,12 +2367,22 @@
...
+    "    new_df = df.copy()\n",
+    "    brier_score = float(0)\n",
+    "\n",
+    "    for column in new_df.columns:\n",
+    "        for correctlabel, probability in zip(correctlabels,  new_df[column]):\n",
+    "            outcome = 1 if correctlabel == column else 0\n",
+    "            brier_score += (probability - outcome) ** 2\n",
+    "\n",
+    "    return brier_score/len(new_df)\n"
    ]
```
[the remark this answers is not in the document]
Comment (Atom0713) [reply Assigment_1/Assignment 1.ipynb:2381]: Is it just a different version or is there a mistake?

**Pull request merged**